File: leetcode-daily/09-January/Smallest Subtree with all the Deepest Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public List<TreeNode> bfs(TreeNode root){

        List<TreeNode> ans = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);

        while(!q.isEmpty()){
            int sz = q.size();

            for(int i=0;i<sz;i++){
                TreeNode curr = q.poll();
                ans.add(curr);
                if(curr.left!=null)q.offer(curr.left);
                if(curr.right!=null)q.offer(curr.right);
            }
            if(!q.isEmpty())ans.clear();
        }
        return ans;
    }
    public TreeNode lca(TreeNode root,TreeNode p,TreeNode q){
        if(root==null || p==root || q==root)return root;

        TreeNode l = lca(root.left,p,q);
        TreeNode r = lca(root.right,p,q);

        if(l==null)return r;
        else if(r==null)return l;
        return root;
    }
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        
        List<TreeNode> leaf = bfs(root);

        TreeNode curr = leaf.get(0);
        curr=lca(root,curr,leaf.get(leaf.size()-1));
        return curr;
    }
}

File: leetcode-daily/09-January/Smallest Subtree with all the Deepest Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> leavesdeep(TreeNode* root){
        
        queue<TreeNode* > q;
        
        q.push(root);
        
        while(true){
            int sz = q.size();
            vector<TreeNode*> row;
            for(int i=0;i<sz;i++){
                root = q.front();
                q.pop();
                row.push_back(root);
                if(root->left){
                    q.push(root->left);
                }
                if(root->right){
                    q.push(root->right);
                }
            }
            if(q.empty())return row;
        }
    }
    TreeNode* LCA(TreeNode* root,TreeNode* m,TreeNode* n){
        
        if(root==NULL || m==root || n==root)return root;
        
        TreeNode* l = LCA(root->left,m,n);
        TreeNode* r = LCA(root->right,m,n);
        
        if(!l)return r;
        else if(!r)return l;
        else return root;
        
    }
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        
        if(root==NULL)return NULL;
        
        vector<TreeNode*> leaf = leavesdeep(root);
        
        TreeNode* ans = leaf[0];
        ans = LCA(root,ans,leaf[leaf.size()-1]);
        
        return ans;
    }
};